### PR TITLE
command/cp&rm: add `--raw` flag

### LINKED ISSUE #235: disable wildcard and use "*" as literal?
The S3 objects I'm going to download has "\*" in object key. When I use `cp` to download, `s5cmd` expands it and download multiple objects more than the particular object. Is there a way to escape wildcard "\*" or disable wildcard? Thanks.

### PR DESCRIPTION
**Pull request by ocakhasan**
- command/cp&rm:  new flag `--raw` which disables the wildcard operations.
- also changed the `rb` command's helper text to correct form, based on [this commit](https://github.com/peak/s5cmd/commit/68f797b0de6b26f311076012928a9aa3aaad2713).

resolves #235

### CODE REVIEW

File: command/cp.go [hunk 1]
```diff
@@ -280,7 +288,8 @@ func (c Copy) Run(ctx context.Context) error {
 		}
 	}()
 
-	isBatch := srcurl.HasGlob()
+	// use raw flag to handle glob operations.
```
Comment (aykutfarsak) [inline command/cp.go:291]:
```suggestion
	// use --raw flag to prevent glob operations.
```
or you can remove the comment entirely.

File: command/expand.go [hunk 2]
```diff
@@ -16,22 +16,28 @@ func expandSource(
...
-	// if the source is local, we send a Stat call to know if  we have
-	// directory or file to walk. For remote storage, we don't want to send
-	// Stat since it doesn't have any folder semantics.
-	if !srcurl.HasGlob() && !srcurl.IsRemote() {
-		obj, err := client.Stat(ctx, srcurl)
-		if err != nil {
-			return nil, err
+
+	// If wildcard operations are disabled (raw flag is set), then program deals with 1 file.
+	// Therefore, there is no need to check the directory operations.
```
Comment (aykutfarsak) [inline command/expand.go:23]:
```suggestion
	// if wildcard operations are disabled with the raw flag, then the program deals with 1 file.
	// therefore, there is no need to check if it is a directory.
```
I am not sure about the comment but let's not start with capital letters.

File: command/expand_test.go [hunk 3]
```diff
@@ -328,6 +329,86 @@ func TestExpandSource_Do_Not_Follow_Symlinks(t *testing.T) {
 	assert.Equal(t, []string{workdirJoin}, expected)
 }
 
+func TestExpandSourceWithRawFlagTrue(t *testing.T) {
```
Comment (aykutfarsak) [inline command/expand_test.go:332]:
```suggestion
func TestExpandSourceWithRawFlag(t *testing.T) {
```

File: command/expand_test.go [hunk 4]
```diff
@@ -328,6 +329,86 @@ func TestExpandSource_Do_Not_Follow_Symlinks(t *testing.T) {
...
+	assert.Equal(t, len(ch), 1)
+	var expected []string
+	for obj := range ch {
+		expected = append(expected, obj.URL.Absolute())
+	}
+	workdirJoin := filepath.ToSlash(workdir.Join("a/file*.txt"))
+	assert.Equal(t, []string{workdirJoin}, expected)
+}
+
+func TestExpandSourceWithRawFlagFalse(t *testing.T) {
```
Comment (aykutfarsak) [inline command/expand_test.go:370]:
```suggestion
func TestExpandSourceWithoutRawFlag(t *testing.T) {
```

Comment (ocakhasan) [reply command/expand_test.go]: changed to `TestExpandSourceWithoutRawFlag`

File: command/expand_test.go (hunk 3 above)
Comment (ocakhasan) [reply command/expand_test.go]: changed to `TestExpandSourceWithRawFlag`.

File: command/expand.go (hunk 2 above)
Comment (ocakhasan) [reply command/expand.go]: fixed the comments as you suggested.

File: command/cp.go (hunk 1 above)
Comment (ocakhasan) [reply command/cp.go]: changed the comment as you suggested.

File: command/expand.go [hunk 5]
```diff
@@ -16,22 +16,28 @@ func expandSource(
 	client storage.Storage,
 	followSymlinks bool,
 	srcurl *url.URL,
+	isRaw bool,
```
Comment (ilkinulas) [inline command/expand.go:19]: There is another way to implement this `--raw` flag: 
Instead of introducing a new boolean parameter to this function we can create a new function that just returns a channel with single object. `cp` and `rm` commands could check `--raw` flag and decide to call either `expandSource` or this new function (`rawSource` for example).

```go
func rawSource(srcurl *url.URL, followSymlinks bool) <-chan *storage.Object {
	ch := make(chan *storage.Object, 1)
	if storage.ShouldProcessUrl(srcurl, followSymlinks) {
		ch <- &storage.Object{URL: srcurl}
	}
	close(ch)
	return ch
}
```

This implementation has following advantages:
- `expandSource` function does what it's name says
- No need to update `expandSource` tests

One drawback that comes to my mind is `cp` and `rm` commands will have one additional check for `--raw` flag, but it would not be more complicated that updating `expandSource` function. 

What do you think?

Comment (ocakhasan) [reply command/expand.go:19]: I think this approach is better than with parameter, because in the other [PR](https://github.com/peak/s5cmd/pull/324), I also had to add another parameter (`exclude`) to `expandSource`, and if I add parameters to `expandSource` like this, then it would be very hard to follow. I will change to new function as you suggested.

File: command/expand.go [hunk 6]
```diff
@@ -89,3 +90,25 @@ func expandSources(
...
+	ch := make(chan *storage.Object, 1)
+	if storage.ShouldProcessUrl(srcurl, followSymlinks) {
+		ch <- &storage.Object{URL: srcurl}
+	}
+	close(ch)
+	return ch
+}
+
+// raw source returns the only object with given path.
+func rawSourceUrls(srcurls []*url.URL, followSymlinks bool) <-chan *storage.Object {
```
Comment (ilkinulas) [inline command/expand.go:105]: Let's try merging `rawSource` and `rawSources` functions into a single variadic function.

```suggestion
func rawSource(followSymlinks bool, srcurls ...*url.URL) <-chan *storage.Object {
```

We can use this function in both cp/rm commands and `expandSource`  function.

Comment (ocakhasan) [reply command/expand.go]: created a new single variadic function.

File: command/expand.go [hunk 7]
```diff
@@ -89,3 +90,15 @@ func expandSources(
 
 	return ch
 }
+
+// raw source returns the only object with given path.
+func rawSource(followSymlinks bool, srcurls ...*url.URL) <-chan *storage.Object {
```
Comment (ilkinulas) [inline command/expand.go:95]: Let's use `rawSource`  to eliminate code duplication in  `expandSource`.

`expand.go` line 38
```go
return rawSource(followSymlinks, srcurl), nil
```

File: command/rm.go [hunk 8]
```diff
@@ -85,7 +95,10 @@ func (d Delete) Run(ctx context.Context) error {
 		return err
 	}
 
-	objChan := expandSources(ctx, client, false, srcurls...)
+	objChan := rawSource(false, srcurls...)
```
Comment (ilkinulas) [inline command/rm.go:98]: Let's revert this raw check to make it more readable.

```go
declare objch channel
if raw {
	objch = rawSource ...
} else {
	objch, err = expandSource ..
}
```

File: command/expand.go (hunk 7 above)
Comment (ocakhasan) [reply command/expand.go:90]: changed to `return rawSource(followSymlinks, srcurl), nil` to avoid code duplication.

File: command/rm.go (hunk 8 above)
Comment (ocakhasan) [reply command/rm.go]: changed as you suggested.

File: e2e/cp_test.go [hunk 9]
```diff
@@ -2926,3 +2926,311 @@ func TestCopyS3ToDirDryRun(t *testing.T) {
...
+
+			// assert filesystem
+			expected := fs.Expected(t, tc.src...)
+			assert.Assert(t, fs.Equal(workdir.Path(), expected))
+		})
+	}
+
+}
+
+// When folder is uploaded with --raw flag, it gives an error.
```
Comment (sonmezonur) [inline e2e/cp_test.go:3049]: Why? I think it should copy files in `a*` directory in raw mode.

**Review by sonmezonur — COMMENTED** [review]
Thanks for the PR :) I tested your branch and it works perfectly fine but I think we can still make some improvements:

1 ) I think we should allow glob characters for destination URL if `--raw` flag is given. Currently we are getting error:
```
❯ ./s5cmd cp --raw test 's3://bigtmp/prefix/test*'
ERROR "cp test s3://bigtmp/prefix/test*": target "s3://bigtmp/prefix/test*" can not contain glob characters
```
2) In raw mode, we are still parsing wildcard strings and compiling regexes even if they are not used. (url.New()  makes lots of parsing and regex calculations([ref](https://github.com/peak/s5cmd/blob/master/storage/url/url.go#L103))). IMO, URL package should know if it is a raw mode and should not make unnecessary/expensive calculations. I prototyped URL package to get optional `URLMode` parameter and skip parsing if `URLMode` is raw ([Prototype](https://github.com/peak/s5cmd/blob/url_mode_prototype/storage/url/url.go)). If we prefer using this method, we don't need to update `command/expand.go` anymore.  If we decide to add raw mode to other commands (like ls, cat, etc.) in near future, it will be quite easy since URL package is used by all other commands (while expandSource is only used by cp and rm). Does this make sense to you?

3) Could you add tests for remote->remote copy operations? As far as I can see, we have tests for local->remote and remote-> local but could not see remote->remote.

File: e2e/cp_test.go (hunk 9 above)
Comment (ocakhasan) [reply e2e/cp_test.go]: Thanks for the feedback. I changed as you suggested and here is my thoughts.

1. Now we can have wildcard in the destination in `--raw` mode. 

2. Changing URL package is simpler than adding new functions or changing `expandSource` function. It is working good and adding `--raw` flag to other commands is much more easier now. 

3. Added tests for remote -> remote. 

Also, local directory `a*` is uploaded now without uploading other folders matching `a*` pattern.

File: storage/url/url.go [hunk 10]
```diff
@@ -45,10 +44,26 @@ type URL struct {
 	relativePath string
 	filter       string
 	filterRegex  *regexp.Regexp
+	mode         URLMode
+}
+
+type URLMode int
```
Comment (ilkinulas) [inline storage/url/url.go:50]: using a bool field instead of URLMode type will simplify checks below:
```go
    var urlMode url.URLMode
    if c.raw {
        urlMode = 1
    }
    srcurl, err := url.New(c.src, url.WithMode(urlMode))
```

Above sample can be written as this if we use a bool field.
```go
srcurl, err := url.New(c.src, url.WithMode(c.raw))
```

File: command/expand.go (hunk 7 above)
Comment (ilkinulas) [reply command/expand.go:90]: Please revert `expand.go`, implementation is moved to `url`.

File: storage/url/url.go (hunk 10 above)
Comment (ocakhasan) [reply storage/url/url.go]: changed as you suggested.

File: command/expand_test.go [hunk 11]
```diff
@@ -328,6 +329,70 @@ func TestExpandSource_Do_Not_Follow_Symlinks(t *testing.T) {
 	assert.Equal(t, []string{workdirJoin}, expected)
 }
 
+func TestRawSourceWithoutSymLinks(t *testing.T) {
```
Comment (ilkinulas) [inline command/expand_test.go:332]: Do we still need these tests?

Comment (ocakhasan) [reply command/expand_test.go:332]: We use `rawSource` method in `expandSource` but it was due to eliminate code duplication. With the change in url package there is no need to use `rawSource` method, therefore I can remove it and change the `expandSource` method to the normal version.

File: command/cp.go [hunk 12]
```diff
@@ -750,7 +759,7 @@ func validateCopyCommand(c *cli.Context) error {
 	}
 
 	// wildcard destination doesn't mean anything
-	if dsturl.HasGlob() {
+	if !c.Bool("raw") && dsturl.HasGlob() {
```
[the remark this answers is not in the document]
Comment (sonmezonur) [reply command/cp.go:762]: Also, please create URLs with `urlMode` option in validateCopyCommand ([1](https://github.com/peak/s5cmd/blob/ae2bcea77d920afcfef1444a7749d51fe20f44ca/command/cp.go#L751), [2](https://github.com/peak/s5cmd/blob/ae2bcea77d920afcfef1444a7749d51fe20f44ca/command/cp.go#L756)). 

Validation fails if target URL has glob in --raw mode: 
```
❯ ./s5cmd cp --raw 'test*' 's3://bigtmp/test*'
ERROR "cp test* s3://bigtmp/test*": target "s3://bigtmp/test*" must be a bucket or a prefix
```

After updating with urlMode:
```
❯ ./s5cmd cp --raw 'test*' 's3://bigtmp/test*'
cp test* s3://bigtmp/test*
```

Also, we can add simple test for this scenario.

File: storage/url/url.go [hunk 13]
```diff
@@ -291,7 +312,7 @@ func (u *URL) MarshalJSON() ([]byte, error) {
 
 // HasGlob reports whether if a string contains any wildcard chars.
 func (u *URL) HasGlob() bool {
```
Comment (sonmezonur) [inline storage/url/url.go:314]:
```suggestion
func (u *URL) IsWildcard() bool {
```
At this stage, I think we can rename this function since it also checks raw mode. Not sure about naming :D

File: command/expand.go [hunk 14]
```diff
@@ -17,6 +17,7 @@ func expandSource(
 	followSymlinks bool,
 	srcurl *url.URL,
 ) (<-chan *storage.Object, error) {
+
```
[the remark this answers is not in the document]
Comment (ocakhasan) [reply command/expand.go]: deleted empty line.

File: command/cp.go [hunk 15]
```diff
@@ -253,13 +258,15 @@ increase the open file limit or try to decrease the number of workers with
 
 // Run starts copying given source objects to destination.
 func (c Copy) Run(ctx context.Context) error {
-	srcurl, err := url.New(c.src)
+
```
[the remark this answers is not in the document]
Comment (ocakhasan) [reply command/cp.go]: deleted empty line.

File: storage/url/url.go [hunk 16]
```diff
@@ -1,4 +1,3 @@
-// Package url abstracts local and remote file URLs.
```
[the remark this answers is not in the document]
Comment (ocakhasan) [reply storage/url/url.go]: added package comment.

File: command/cp.go [hunk 17]
```diff
@@ -253,13 +258,15 @@ increase the open file limit or try to decrease the number of workers with
...
+	srcurl, err := url.New(c.src, url.WithRaw(c.raw))
 	if err != nil {
 		printError(c.fullCommand, c.op, err)
 		return err
 	}
 
 	dsturl, err := url.New(c.dst)
```
[the remark this answers is not in the document]
Comment (ocakhasan) [reply command/cp.go]: added `url.WithRaw(c.raw)` to url creation.

File: command/cp.go [hunk 18]
```diff
@@ -302,7 +311,8 @@ func (c Copy) Run(ctx context.Context) error {
 		}
 	}()
 
-	isBatch := srcurl.HasGlob()
+	// use --raw flag to prevent glob operations.
+	isBatch := !c.raw && srcurl.HasGlob()
```
[the remark this answers is not in the document]
Comment (ocakhasan) [reply command/cp.go]: changed as you suggested.

File: command/cp.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (ocakhasan) [reply command/cp.go]: changed as you suggested. and added a new test case in `cp_test.go`.

File: storage/url/url.go (hunk 13 above)
Comment (ocakhasan) [reply storage/url/url.go:315]: I am also not sure about the name, so if we can agree on a name, then I will change it 😄

File: command/rm.go [hunk 19]
```diff
@@ -65,14 +74,16 @@ type Delete struct {
...
+	raw         bool
 
 	// storage options
 	storageOpts storage.Options
 }
 
 // Run remove given sources.
 func (d Delete) Run(ctx context.Context) error {
-	srcurls, err := newURLs(d.src...)
+
```
[the remark this answers is not in the document]
Comment (ocakhasan) [reply command/rm.go]: deleted the empty line.

File: storage/url/url_test.go [hunk 20]
```diff
@@ -402,3 +402,33 @@ func TestURLIsBucket(t *testing.T) {
 		}
 	}
 }
+
+func TestURLWithMode(t *testing.T) {
+	tests := []struct {
+		input          string
+		urlMode        bool
```
Comment (ilkinulas) [inline storage/url/url_test.go:409]:
```suggestion
		raw        bool
```

Comment (ocakhasan) [reply storage/url/url_test.go]: changed to `raw`.

File: e2e/cp_test.go [hunk 21]
```diff
@@ -2926,3 +2926,460 @@ func TestCopyS3ToDirDryRun(t *testing.T) {
...
+	result := icmd.RunCmd(cmd)
+
+	result.Assert(t, icmd.Success)
+
+	assertLines(t, result.Stdout(), map[int]compareFunc{
+		0: equals("cp %v/file.txt %v/a*/file.txt", srcpath, dstpath),
+		1: equals("cp %v/file1.txt %v/a*/file1.txt", srcpath, dstpath),
+	}, sortInput(true))
+
+	// assert no files are in S3
```
Comment (ilkinulas) [inline e2e/cp_test.go:3093]: we do not need this comment
```suggestion
```

Comment (ocakhasan) [reply e2e/cp_test.go]: deleted the comment line.

File: storage/url/url.go (hunk 13 above)
Comment (sonmezonur) [reply storage/url/url.go:315]: My choice is `IsWildcard()`

File: e2e/cp_test.go [hunk 22]
```diff
@@ -2926,3 +2926,459 @@ func TestCopyS3ToDirDryRun(t *testing.T) {
...
+		},
+		{
+			name:       "cp  a*/file.txt s3://bucket/",
+			src:        []string{"a*/file.txt", "a*/file1.txt", "a*/file2.txt"},
+			wantedFile: "a*/file.txt",
+			rawFlag:    "--raw",
+			expectedFiles: []fs.PathOp{
+				fs.WithFile("file.txt", fileContent),
+			},
+		},
```
Comment (ilkinulas) [inline e2e/cp_test.go:3159]:
```suggestion
		{
			name:       "cp  a*/file.txt s3://bucket/",
			src:        []string{"a*/file*.txt", "a*b/file1.txt", "a*c/file2.txt"},
			wantedFile: "a*/file*.txt",
			rawFlag:    "--raw",
			expectedFiles: []fs.PathOp{
				fs.WithFile("file*.txt", fileContent),
			},
		},
```
This tests raw mode better, what do you think?

File: storage/url/url.go (hunk 13 above)
Comment (ocakhasan) [reply storage/url/url.go]: `HasGlob()` is changed to `IsWildcard()`

File: e2e/cp_test.go (hunk 22 above)
Comment (ocakhasan) [reply e2e/cp_test.go]: I think you are right. I tried to change other tests also.

File: e2e/cp_test.go [hunk 23]
```diff
@@ -2926,3 +2926,424 @@ func TestCopyS3ToDirDryRun(t *testing.T) {
...
+				fs.WithFile("file*2.txt", "content"),
+			},
+			wantedFile:       "file*.txt",
+			dst:              "s3://bucket/",
+			expectedFiles:    []string{"file*.txt", "file*1.txt", "file*file.txt", "file*2.txt"},
+			nonExpectedFiles: []string{},
+			rawFlag:          "",
+		},
+		{
+			name: "cp  a*/file.txt s3://bucket/",
```
Comment (ilkinulas) [inline e2e/cp_test.go:2979]:
```suggestion
			name: "cp  a*/file*.txt s3://bucket/",
```

File: e2e/cp_test.go [hunk 24]
```diff
@@ -2926,3 +2926,424 @@ func TestCopyS3ToDirDryRun(t *testing.T) {
...
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+
+			s3client, s5cmd, cleanup := setup(t)
+			defer cleanup()
+
+			createBucket(t, s3client, bucket)
+
+			workdir := fs.NewDir(t, tc.name, tc.src...)
```
Comment (ilkinulas) [inline e2e/cp_test.go:3012]:
```suggestion
			workdir := fs.NewDir(t, "copy-raw-test", tc.src...)
```

File: e2e/cp_test.go (hunk 23 above)
Comment (ocakhasan) [reply e2e/cp_test.go]: changed as you suggested.

File: e2e/cp_test.go (hunk 24 above)
Comment (ocakhasan) [reply e2e/cp_test.go]: changed as you suggested.

**Review by ilkinulas — APPROVED** [review]

File: command/cp.go [hunk 25]
```diff
@@ -302,7 +309,8 @@ func (c Copy) Run(ctx context.Context) error {
 		}
 	}()
 
-	isBatch := srcurl.HasGlob()
+	// use --raw flag to prevent glob operations.
```
Comment (sonmezonur) [inline command/cp.go:312]: you can remove this comment since we don't check --raw here anymore.

Comment (ocakhasan) [reply command/cp.go]: deleted comment line.

**Review by sonmezonur — APPROVED** [review]
